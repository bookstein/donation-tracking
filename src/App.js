import React, { Component } from 'react'
import './App.css'
import ProfileContainer from './containers/profileContainer'

class App extends Component {
  render() {
    return <ProfileContainer />
  }
}

export default App
