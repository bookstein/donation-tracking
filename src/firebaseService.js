import fb from './firebase'

export const listenForUpdates = (refName, cb) => {
  let myRef = fb
    .database()
    .ref(refName)
    .orderByKey()

  myRef.on('child_added', cb)
  myRef.on('child_removed', cb)
}

export const pushToDatabase = (refName, data) => {
  const myRef = fb
    .database()
    .ref(refName)
    .push()
  myRef.set(data)
  // for some reason I can only use `set`
  // `push` seems to append to the last item in the collection
  console.log('view your new data at ', myRef.toString())
}

export const removeFromDatabase = (refName, key) => {
  const myRef = fb.database().ref(refName)
  myRef
    .child(key)
    .remove()
    .then(function() {
      console.log('Remove succeeded.')
    })
    .catch(function(error) {
      console.log('Remove failed: ' + error.message)
    })
}
