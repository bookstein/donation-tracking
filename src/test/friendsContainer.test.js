import React from 'react'
import ReactDOM from 'react-dom'
import App from '../App'
import FriendsContainer from '../containers/friendsContainer'

// get testing tools
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'
import * as firebaseService from '../firebaseService'

jest.mock('../firebaseService', () => {
  return {
    listenForUpdates: jest.fn(),
    pushToDatabase: jest.fn(),
  }
})

// the "edges" of friends container
jest.mock('../components/friend', () => {
  return () => <div>Friend</div>
})

const SNAPSHOT_FROM_FIREBASE = {
  key: 'snapshotKey',
  val: () => ({
    exerciseName: 'myCoolExercise',
    exerciseTag: 'legs',
  }),
}
const SNAPSHOT_AS_STATE = {
  id: 'snapshotKey',
  text: { exerciseName: 'myCoolExercise', exerciseTag: 'legs' },
}

/*
HAPPY PATH:
  component renders as we expect
*/
it('renders the happy path', () => {
  const component = renderer.create(<FriendsContainer />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

/*
BEHAVIOR:
  what behavior does the container itself have?
    1. submitting forms
    2. loading data
*/
it('sets state with initial list of friends', done => {
  // NOTE: mock needs to match the behavior of the original
  // listenForUpdates('friends', this.buildExerciseList) -> must pass a fake snapshot into cb
  firebaseService.listenForUpdates.mockImplementation((_, cb) =>
    cb(SNAPSHOT_FROM_FIREBASE),
  )
  const wrapper = mount(<FriendsContainer />)
  wrapper.update()
  setImmediate(() => {
    const friends = wrapper.state('friends')
    expect(friends).toHaveLength(1)
    expect(friends[0]).toMatchObject(SNAPSHOT_AS_STATE)
  })
  done()
})
it('sends new friends to Firebase on form submit', () => {
  // note: this is not really a unit test, more integration test
  firebaseService.listenForUpdates.mockImplementation((_, cb) =>
    cb(SNAPSHOT_FROM_FIREBASE),
  )
  const wrapper = mount(<FriendsContainer />)
  wrapper.find('.AddFriendForm__Submit').simulate('submit')
  expect(firebaseService.pushToDatabase.mock.calls).toHaveLength(1)
})

/*
EDGE CASES
  what are our container's edges?
    1. calls to external services (Firebase)
      - call fails
      - call returns no data

*/

it('renders an error message if Firebase fails', done => {
  // TODO: find out what errors from Firebase realistically look like
  done()
})
