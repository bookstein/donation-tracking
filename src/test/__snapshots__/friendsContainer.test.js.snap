// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the happy path 1`] = `
<div
  className="FriendsContainer"
>
  <div
    className="LeftColumn"
  >
    <img
      alt="socialjusticefund.org"
      height="270"
      src="sjf.png"
      width="214"
    />
    <div
      className="AddFriendForm"
    >
      <form
        onSubmit={[Function]}
      >
        <h3>
          Add New Friend
        </h3>
        <div
          className="AddFriendForm__FormGroup"
        >
          <label>
            Name
          </label>
          <input
            className="AddFriendForm__Input"
            name="friendName"
            onChange={[Function]}
            type="text"
            value={undefined}
          />
        </div>
        <div
          className="AddFriendForm__FormGroup"
        >
          <label>
            Passionate about:
          </label>
          <input
            className="AddFriendForm__Input"
            name="friendTag"
            onChange={[Function]}
            type="text"
            value={undefined}
          />
        </div>
        <input
          className="AddFriendForm__Submit"
          type="submit"
        />
      </form>
    </div>
  </div>
  <div
    className="RightColumn"
  >
    <h2>
      Ask Friends to Donate
    </h2>
  </div>
</div>
`;
